Dockerfile

```diff
@@ -1,20 +1,42 @@
 # Initial Docker support thanks to xinyifly
+# Optimisation performed by wejrox
 
 #
-# Build stage
+# Cosmic JAR creation stage
 #
-FROM maven:3.6.3-jdk-8 AS build
-COPY src /home/app/src
-COPY pom.xml /home/app
-RUN mvn -f /home/app/pom.xml clean package
+FROM maven:3.6.3-jdk-8 AS jar
+
+# Build in a separated location which won't have permissions issues.
+WORKDIR /opt/cosmic
+# Any changes to the pom will affect the entire build, so it should be copied first.
+COPY pom.xml ./pom.xml
+# Grab all the dependencies listed in the pom early, since it prevents changes to source code from requiring a complete re-download.
+# Skip compiling tests since we don't want all the dependecies to be downloaded.
+RUN mvn -f ./pom.xml clean dependency:go-offline -Dmaven.test.skip -T 1C
+# Source code changes may not change dependencies, so it can go last.
+# Skip compiling tests since we don't want all the dependecies to be downloaded for plugins.
+COPY src ./src
+RUN mvn -f ./pom.xml clean package -Dmaven.test.skip -T 1C
```

Comment:
Do you really need to specify `-T 1C`? I had to look it up, and it's to set the thread count to 1, which seems obscure to me. The simpler the better.

Reply:
It's not required, but the option tells maven to run multithreaded with one thread per core that is available, which speeds up module compilation and package time.

Reply:
That's reassuring; PR approved.